panther_battery/include/panther_battery/battery_publisher.hpp

```diff
@@ -0,0 +1,182 @@
+}
+
+inline BatteryPublisher::~BatteryPublisher() { battery_pub_.reset(); }
+
+inline void BatteryPublisher::PublishUnknown(rclcpp::Time header_stamp)
+{
+  battery_voltage_ma_->Reset();
+  battery_temp_ma_->Reset();
+  battery_current_ma_->Reset();
+  battery_charge_ma_->Reset();
+
+  auto battery_msg = BatteryStateMsg();
+  battery_msg.header.stamp = header_stamp;
+  battery_msg.voltage = std::numeric_limits<double>::quiet_NaN();
+  battery_msg.temperature = std::numeric_limits<double>::quiet_NaN();
+  battery_msg.current = std::numeric_limits<double>::quiet_NaN();
+  battery_msg.percentage = std::numeric_limits<double>::quiet_NaN();
+  battery_msg.capacity = std::numeric_limits<double>::quiet_NaN();
+  battery_msg.design_capacity = battery_designed_capacity_;
+  battery_msg.charge = std::numeric_limits<double>::quiet_NaN();
+  battery_msg.cell_voltage = std::vector<float>(10, std::numeric_limits<float>::quiet_NaN());
+  battery_msg.cell_temperature = std::vector<float>(10, std::numeric_limits<float>::quiet_NaN());
+  battery_msg.power_supply_status = BatteryStateMsg::POWER_SUPPLY_STATUS_UNKNOWN;
+  battery_msg.power_supply_health = BatteryStateMsg::POWER_SUPPLY_HEALTH_UNKNOWN;
+  battery_msg.power_supply_technology = BatteryStateMsg::POWER_SUPPLY_TECHNOLOGY_LIPO;
+  battery_msg.present = true;
+  battery_msg.location = location_;
+
+  battery_pub_->publish(battery_msg);
+}
+
+inline void BatteryPublisher::Publish(
+  rclcpp::Time header_stamp, const float & V_bat, const float & temp_bat, const float & I_bat,
+  const float & I_charge, const bool & charger_connected)
+{
+  battery_voltage_ma_->Roll(V_bat);
+  battery_temp_ma_->Roll(temp_bat);
+  battery_current_ma_->Roll(I_bat);
+  battery_charge_ma_->Roll(I_charge);
+
+  auto V_bat_mean = battery_voltage_ma_->GetAverage();
+  auto temp_bat_mean = battery_temp_ma_->GetAverage();
+  auto I_bat_mean = battery_current_ma_->GetAverage();
+  auto I_charge_mean = battery_charge_ma_->GetAverage();
+
+  auto battery_msg = BatteryStateMsg();
+  battery_msg.header.stamp = header_stamp;
+  battery_msg.voltage = V_bat_mean;
+  battery_msg.temperature = temp_bat_mean;
+  battery_msg.current = I_bat_mean;
+  battery_msg.percentage =
+    std::clamp((V_bat_mean - V_bat_min_) / (V_bat_full_ - V_bat_min_), 0.0, 1.0);
+  battery_msg.capacity = std::numeric_limits<double>::quiet_NaN();
+  battery_msg.design_capacity = battery_designed_capacity_;
+  battery_msg.charge = battery_msg.percentage * battery_msg.design_capacity;
+  battery_msg.cell_voltage = std::vector<float>(10, std::numeric_limits<float>::quiet_NaN());
+  battery_msg.cell_temperature = std::vector<float>(10, std::numeric_limits<float>::quiet_NaN());
+  battery_msg.power_supply_technology = BatteryStateMsg::POWER_SUPPLY_TECHNOLOGY_LIPO;
+  battery_msg.present = true;
+  battery_msg.location = location_;
+
+  // check battery status
+  if (charger_connected) {
+    if (battery_msg.percentage == 1.0) {
```

Comment:
This will never be true. In c++ comparing floats has to be done this way:
```suggestion
    if (fabs(battery_msg.percentage - 1.0f) < std::numeric_limits<float>::epsilon()) {
```

Reply:
Done

Reply:
Done